cdap-event-reader-spi/src/main/java/io/cdap/cdap/spi/events/EventReader.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.spi.events;
+
+import java.util.List;
+
+/**
+ * {@link EventReader} Interface for listening for events
+ */
+public interface EventReader<T extends Event> extends AutoCloseable {
+  /**
+   * Method to initialize EventReader.
+   *
+   * @param context Configuration properties of reader
+   */
+  void initialize(EventReaderContext context);
+
+  /**
+   * Pull messages if available.
+   *
+   * @param maxMessages maximum messages to pull
+   * @return Result containing events
+   */
+  EventResult<T> pull(int maxMessages);
+
+  /**
+   * Returns the identifier for this reader.
+   *
+   * @return String id for the reader
+   */
+  String getId();
```

Comment:
Why is this part of the interface contract?

Reply:
Following the [EventWriter pattern](https://github.com/cdapio/cdap/blob/7305ff122c8042dbc20d3db948bd156aa6f5a24d/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventWriterExtensionProvider.java#LL50C6-L50C6), the config contains a list of enabled event readers that are identified by this id. If the id of the specific event reader extension is not specified in the config, it is not enabled/loaded in App Fabric

Reply:
Requiring the code to return an ID that has to be unique among all implementation is not a good contract. Typically we have sub-directories under a particular extension directory and the subdir name is the id/name of the extension, hence relying on code to return it unique ID. @greeshmaswaminathan do you know why we had the code to return ID instead of using dir name?

Reply:
Referred from the [MetricsWriter](https://github.com/cdapio/cdap/blob/develop/cdap-watchdog-api/src/main/java/io/cdap/cdap/api/metrics/MetricsWriter.java) interface .  Since the ID was already available used that for enabling different writers for different deployments. Makes sense to get this from directory name, nothing stops two implementations from providing the same ID currently. Created [JIRA](https://cdap.atlassian.net/browse/CDAP-20700) for this.